pkg/events/dependencies/manager.go

```diff
@@ -396,11 +480,141 @@ func (m *Manager) buildProbe(handle probes.Handle, dependent events.ID) error {
+// indicating the event was not removed but switched to a fallback.
+// If no fallbacks are available or all fail, the event is removed and the function returns (true, error),
+// where the error is from the removal or nil if successful.
+func (m *Manager) failEvent(eventID events.ID) (bool, error) {
+	// Check if already processing this event (prevents loops)
+	if _, processing := m.processingEvents[eventID]; processing {
+		return false, fmt.Errorf("event %v is currently being processed", eventID)
+	}
+
+	node := m.getEventNode(eventID)
+	if node == nil {
+		return false, ErrNodeNotFound
+	}
+
+	// Mark as processing
+	m.processingEvents[eventID] = struct{}{}
+	defer delete(m.processingEvents, eventID) // Always cleanup
+
+	return m.failEventNode(node)
+}
+
+func (m *Manager) failEventNode(node *EventNode) (bool, error) {
+	eventName := events.Core.GetDefinitionByID(node.GetID()).GetName()
+	logger.Debugw("Failing event", "event", eventName)
+	// Try fallbacks in a loop until one succeeds or we run out of fallbacks
+	for node.hasMoreFallbacks() {
+		m.removeEventNodeFromDependencies(node)
+		if !node.fallback() {
+			break
+		}
+		logger.Debugw("Switching to fallback", "event", eventName, "fallback number", node.currentFallbackIndex)
+
+		_, err := m.buildEventNode(node)
+		if err == nil {
+			logger.Debugw("Successfully switched to fallback", "event", eventName)
+			return false, nil
+		}
+		logger.Debugw("Failed to switch to fallback", "event", eventName, "fallback number", node.currentFallbackIndex, "error", err)
+	}
+	logger.Debugw("All fallbacks failed, removing event", "event", eventName)
+	return true, m.removeEventNode(node)
+}
+
+// handleEventAddError handles the error of adding a node to the tree.
+// As errors in adding a node to the tree are caused either by dependencies failure or by add-watchers actions,
+// it will determine the handling strategy based on the error type.
+func (m *Manager) handleEventAddError(node *EventNode, err error) error {
+	var cancelErr *ErrNodeAddCancelled
+
+	if errors.As(err, &cancelErr) {
+		// Cancellation: immediate removal, no fallbacks
+		// No need to fail event dependents, as they will fail using event add failure error handling
+		eventName := events.Core.GetDefinitionByID(node.GetID()).GetName()
+		logger.Debugw("Event add cancelled", "event", eventName)
+		m.removeEventNodeFromDependencies(node)
+		m.removeNode(node)
+		return err
+	}
+	// Failure: try fallbacks
+	removed, failEventErr := m.failEventNode(node)
+	if failEventErr != nil {
+		logger.Errorw("Failed to fail", "error", failEventErr)
+		m.removeEventNodeFromDependencies(node)
+		return err
+	}
+	if removed {
+		// All fallbacks exhausted, event was removed
+		return err
+	}
+	// Fallback succeeded, no error to return
+	return nil
+}
+
+// FailProbe marks a probe as failed and fails all events that depend on it.
+func (m *Manager) FailProbe(handle probes.Handle) error {
```

Comment:
This is similar to RemoveProbe in the logic

Reply:
So I thought about it and the difference between it and remove is that it is marked has failed and so it won't be loaded again. So I decided to keep it.